tensorboard/plugins/hparams/api.py

```diff
@@ -0,0 +1,378 @@
+
+  def __init__(self, min_value=None, max_value=None):
+    if not isinstance(min_value, int):
+      raise TypeError("min_value must be an int: %r" % (min_value,))
+    if not isinstance(max_value, int):
+      raise TypeError("max_value must be an int: %r" % (max_value,))
+    if min_value > max_value:
+      raise ValueError("%r > %r" % (min_value, max_value))
+    self._min_value = min_value
+    self._max_value = max_value
+
+  def __str__(self):
+    return "[%s, %s]" % (self._min_value, self._max_value)
+
+  def __repr__(self):
+    return "IntInterval(%r, %r)" % (self._min_value, self._max_value)
+
+  @property
+  def dtype(self):
+    return int
+
+  @property
+  def min_value(self):
+    return self._min_value
+
+  @property
+  def max_value(self):
+    return self._max_value
+
+  def update_hparam_info(self, hparam_info):
+    hparam_info.type = api_pb2.DATA_TYPE_FLOAT64  # TODO(#1998): Add int dtype.
+    hparam_info.domain_interval.min_value = self._min_value
+    hparam_info.domain_interval.max_value = self._max_value
+
+
+class RealInterval(Domain):
+  """A domain that takes on all real values in a closed interval."""
+
+  def __init__(self, min_value=None, max_value=None):
+    if not isinstance(min_value, float):
+      raise TypeError("min_value must be a float: %r" % (min_value,))
+    if not isinstance(max_value, float):
+      raise TypeError("max_value must be a float: %r" % (max_value,))
+    if min_value > max_value:
+      raise ValueError("%r > %r" % (min_value, max_value))
+    self._min_value = min_value
+    self._max_value = max_value
+
+  def __str__(self):
+    return "[%s, %s]" % (self._min_value, self._max_value)
+
+  def __repr__(self):
+    return "RealInterval(%r, %r)" % (self._min_value, self._max_value)
+
+  @property
+  def dtype(self):
+    return float
+
+  @property
+  def min_value(self):
+    return self._min_value
+
+  @property
+  def max_value(self):
+    return self._max_value
+
+  def update_hparam_info(self, hparam_info):
+    hparam_info.type = api_pb2.DATA_TYPE_FLOAT64
+    hparam_info.domain_interval.min_value = self._min_value
+    hparam_info.domain_interval.max_value = self._max_value
+
+
+class Discrete(Domain):
+  """A domain that takes on a fixed set of values.
+
+  These values may be of any (single) domain type.
+  """
+
+  def __init__(self, values, dtype=None):
+    self._values = list(values)
+    if dtype is None:
+      if self._values:
+        dtype = type(self._values[0])
+      else:
+        raise ValueError("Empty domain with no dtype specified")
+    if dtype not in (int, float, bool, str):
+      raise ValueError("Unknown dtype: %r" % (dtype,))
+    self._dtype = dtype
+    for value in self._values:
+      if not isinstance(value, self._dtype):
+        raise ValueError(
+            "dtype mismatch: not isinstance(%r, %s)"
+            % (value, self._dtype.__name__)
+        )
+    self._values.sort()
```

Comment:
Did not think too much about this: Is this always desired and is this natural for users?

Reply:
It’s a fair question. Logically, a discrete domain is an unordered set,
so it seemed natural to me that the representation should be invariant
under permutation of `values`. But I could be convinced otherwise.

(Note that the frontend already displays discrete axes as sorted,
irrespective of the domain definition.)